crates/adapters/src/static_compile/serialize_with_context.rs

```diff
@@ -0,0 +1,489 @@
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        self.value.serialize_with_context(serializer, self.context)
+    }
+}
+
+impl<C, T> SerializeWithContext<C> for Vec<T>
+where
+    T: SerializeWithContext<C>,
+{
+    fn serialize_with_context<S>(&self, serializer: S, context: &C) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut seq = serializer.serialize_seq(Some(self.len()))?;
+        for element in self {
+            seq.serialize_element(&SerializationContext::new(context, element))?;
+        }
+
+        seq.end()
+    }
+}
+
+impl<C, T> SerializeWithContext<C> for &T
+where
+    T: SerializeWithContext<C>,
+{
+    fn serialize_with_context<S>(&self, serializer: S, context: &C) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        T::serialize_with_context(*self, serializer, context)
+    }
+}
+
+macro_rules! serialize_tuple {
+    ([$num_fields:expr]($(($arg_name:ident, $arg_type:tt)),*)) => {
+        #[allow(unused_variables)]
+        #[allow(dead_code)]
+        #[allow(unused_mut)]
+        impl<C, $($arg_type),*> SerializeWithContext<C> for ($($arg_type),*)
+        where
+            $($arg_type: SerializeWithContext<C>),*
+        {
+            fn serialize_with_context<S>(&self, serializer: S, context: &C) -> Result<S::Ok, S::Error>
+            where
+                S: Serializer,
+            {
+                let ($($arg_name),*) = self;
+                let mut tuple = serializer.serialize_tuple($num_fields)?;
+                $(tuple.serialize_element(&SerializationContext::new(context, $arg_name))?;)*
+                tuple.end()
+            }
+        }
+    }
+}
+
+serialize_tuple!([0]());
+serialize_tuple!([2]((v0, T0), (v1, T1)));
+serialize_tuple!([3]((v0, T0), (v1, T1), (v2, T2)));
+serialize_tuple!([4]((v0, T0), (v1, T1), (v2, T2), (v3, T3)));
+serialize_tuple!([5]((v0, T0), (v1, T1), (v2, T2), (v3, T3), (v4, T4)));
+serialize_tuple!([6](
+    (v0, T0),
+    (v1, T1),
+    (v2, T2),
+    (v3, T3),
+    (v4, T4),
+    (v5, T5)
+));
+serialize_tuple!([7](
+    (v0, T0),
+    (v1, T1),
+    (v2, T2),
+    (v3, T3),
+    (v4, T4),
+    (v5, T5),
+    (v6, T6)
+));
+serialize_tuple!([8](
+    (v0, T0),
+    (v1, T1),
+    (v2, T2),
+    (v3, T3),
+    (v4, T4),
+    (v5, T5),
+    (v6, T6),
+    (v7, T7)
+));
+serialize_tuple!([9](
+    (v0, T0),
+    (v1, T1),
+    (v2, T2),
+    (v3, T3),
+    (v4, T4),
+    (v5, T5),
+    (v6, T6),
+    (v7, T7),
+    (v8, T8)
+));
+serialize_tuple!([10](
+    (v0, T0),
+    (v1, T1),
+    (v2, T2),
+    (v3, T3),
+    (v4, T4),
+    (v5, T5),
+    (v6, T6),
+    (v7, T7),
+    (v8, T8),
+    (v9, T9)
+));
+serialize_tuple!([11](
+    (v0, T0),
+    (v1, T1),
+    (v2, T2),
+    (v3, T3),
+    (v4, T4),
+    (v5, T5),
+    (v6, T6),
+    (v7, T7),
+    (v8, T8),
+    (v9, T9),
+    (v10, T10)
+));
+serialize_tuple!([12](
```

Comment:
why don't we ever need more?

Reply:
we might, but the SQL compiler doesn't generate tuples, so this is not going to cause issues for auto-generated code.